docs: Shareable configs page edits and expansion (#16824)
eslint/eslint: docs/src/extend/shareable-configs.md

---

> To share your `.eslintrc` file, create a **shareable config**. You can publish your shareable config on [npm](https://www.npmjs.com/) so that others can download and use it in their ESLint projects.

Reviewer: Because we will be renaming the config file, let's go generic here.
```suggestion
To share your configuration file, create a **shareable config**. You can publish your shareable config on [npm](https://www.npmjs.com/) so that others can download and use it in their ESLint projects.
```

Reviewer 2: I think we should avoid "file" because shareable configs cannot be used with [`--config`](https://eslint.org/docs/latest/use/command-line-interface#-c---config) and this is rather sharing a configuration object than sharing a file.

 ```suggestion
To share your ESLint configuration, create a **shareable config**. You can publish your shareable config on [npm](https://www.npmjs.com/) so that others can download and use it in their ESLint projects.
```

Thoughts?

Reviewer: Makes sense to me.

};
```

Since the `index.js` file is just JavaScript, you can read these settings from a file or generate them dynamically.

## Publishing a Shareable Config

> Once your shareable config is ready, you can [publish it to npm](https://docs.npmjs.com/getting-started/publishing-npm-packages) to share it with others. We recommend using the `eslint` and `eslintconfig` [keywords](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#keywords) in the `package.json` so others can easily find your module.

Reviewer: ```suggestion
Once your shareable config is ready, you can [publish it to npm](https://docs.npmjs.com/getting-started/publishing-npm-packages) to share it with others. We recommend using the `eslint` and `eslintconfig` [keywords](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#keywords) in the `package.json` file so others can easily find your module.
```

## Using a Shareable Config

> To use a shareable config, include it in the `extends` field of an `.eslintrc` file. For the value, use your module name. For example:

Reviewer 2: I'd keep the "Shareable configs are designed to work with the `extends` feature of `.eslintrc` files." sentence so that it isn't expected that they can be used in other ways. For example, we had a bug report that shareable configs don't work with `--config` (https://github.com/eslint/eslint/issues/15697).

Author: i think the proposed changes are pretty clear that the shareable config should be in the ESLint config file. 

Do you think we should add some sentence like: "You cannot use shareable configs with the ESLint CLI `--config` flag"?

Reviewer 2: > Do you think we should add some sentence like: "You cannot use shareable configs with the ESLint CLI `--config` flag"?

Yes, I think that would be good.

(although, technically it is possible to use a shareable config package with `--config`, e.g. `eslint --config node_modules/eslint-config-airbnb/index.js`)

Reviewer: I think there's still some ambiguity with the current wording. "Include it in the extends field". What is "it"? (We know it's a string, but maybe "it" should "the config name" or something more specific?)

```json
{
    "extends": "eslint-config-myconfig"
}
```

You can also omit the `eslint-config-` and it is automatically assumed by ESLint:

```json
{
    "extends": "myconfig"
}
```

### npm Scoped Modules

npm [scoped modules](https://docs.npmjs.com/misc/scope) are also supported in a number of ways.

You can use the module name:

```json
{
    "extends": "@scope/eslint-config"
}
```

You can also omit the `eslint-config` and it is automatically assumed by ESLint:

```json
{
    "extends": "@scope"
}
```

The module name can also be customized, just note that when using [scoped modules](https://docs.npmjs.com/misc/scope) you cannot omit the `eslint-config-` prefix. Doing so results in package naming conflicts, and thus often in resolution errors. For example, if you have a package named `@scope/eslint-config-myconfig`, the configuration must be specified as:

```json
{
    "extends": "@scope/eslint-config-myconfig"

> }

Reviewer 2: This isn't true, a `@scope/eslint-config-myconfig` package can be used by specifying `"extends": "@scope/myconfig"`. The only case where `"eslint-config"` sequence must appear in the config specifier is when specifying a subpath of a `@scope/eslint-config` package, like in the example on line 150 (`"extends": "@scope/eslint-config/my-special-config"`), because `"@scope/my-special-config"` would be interpreted as `"@scope/eslint-config-my-special-config"`.

Reviewer: You're correct. I think this is a case where the old docs were also incorrect, so probably a good idea to clean this up.

Author: ok, clarifying the text and examples

## Using a Shareable Config

To use a shareable config, include the config name in the `extends` field of a configuration file. For the value, use your module name. For example:

```json
{
    "extends": "eslint-config-myconfig"
}
```

You can also omit the `eslint-config-` and it is automatically assumed by ESLint:

```json
{
    "extends": "myconfig"
}
```

> You cannot use shareable configs with the ESLint CLI [`--config`](../use/command-line-interface#c---config) flag unless you explicitly specify the path to the module (e.g. `eslint --config node_modules/eslint-config-airbnb/index.js`).

Reviewer 2: ```suggestion
You cannot use shareable configs with the ESLint CLI [`--config`](../use/command-line-interface#-c---config) flag.
```

I'd leave just "you cannot" because this is not the intended use of shareable configs, and depends on internal structures of shareable config packages so we shouldn't recommend it in any way.
